FirebaseSessions/Sources/FirebaseSessions.swift

```diff
@@ -89,23 +98,104 @@ protocol SessionsProvider {
-      // Generate a session start event only when session data collection is enabled and if the session is allowed to dispatch events
-      if self.settings.sessionsEnabled, session.shouldDispatchEvents {
-        let event = SessionStartEvent(sessionInfo: session, appInfo: self.appInfo)
-        DispatchQueue.global().async {
-          self.coordinator.attemptLoggingSessionStart(event: event) { result in
-          }
+      // Generating a Session ID early is important as Subscriber
+      // SDKs will need to read it immediately upon registration.
+      let sessionInfo = self.sessionGenerator.generateNewSession()
+
+      // Post a notification so subscriber SDKs can get an updated Session ID
+      self.notificationCenter.post(name: Sessions.SessionIDChangedNotificationName,
+                                   object: nil)
+
+      let event = SessionStartEvent(sessionInfo: sessionInfo, appInfo: self.appInfo)
+
+      // Wait until all subscriber promises have been fulfilled before
+      // doing any data collection.
+      all(self.subscriberPromises.values).then(on: .global(qos: .background)) { _ in
+        guard self.isAnyDataCollectionEnabled else {
+          Logger
+            .logDebug(
+              "Data Collection is disabled for all subscribers. Skipping this Session Event"
+            )
+          return
+        }
+
+        Logger.logDebug("Data Collection is enabled for at least one Subscriber")
+
+        // Fetch settings if they have expired. This must happen after the check for
+        // data collection because it uses the network, but it must happen before the
+        // check for sessionsEnabled from Settings because otherwise we would permanently
+        // turn off the Sessions SDK when we disabled it.
+        self.settings.updateSettings()
+
+        self.addEventDataCollectionState(event: event)
+
+        if !(self.settings.sessionsEnabled && sessionInfo.shouldDispatchEvents) {
+          Logger
+            .logDebug(
+              "Session Event logging is disabled sessionsEnabled: \(self.settings.sessionsEnabled), shouldDispatchEvents: \(sessionInfo.shouldDispatchEvents)"
+            )
+          return
+        }
+
+        self.coordinator.attemptLoggingSessionStart(event: event) { result in
         }
-      } else {
-        Logger.logDebug("Session logging is disabled by configuration settings.")
       }
     }
   }
 
+  // MARK: - Data Collection
+
+  var isAnyDataCollectionEnabled: Bool {
+    for subscriber in subscribers {
+      if subscriber.isDataCollectionEnabled {
+        return true
+      }
+    }
+    return false
+  }
+
+  func addEventDataCollectionState(event: SessionStartEvent) {
+    subscribers.forEach { subscriber in
+      event.set(subscriber: subscriber.sessionsSubscriberName,
+                isDataCollectionEnabled: subscriber.isDataCollectionEnabled)
+    }
+  }
+
+  // MARK: - SessionsProvider
+
+  var currentSessionPayload: SessionPayload {
```

Comment:
Can we return NIL instead of empty string?

Reply:
I wanted to return the object in all cases to prevent crashes in the product SDKs. If the sessionID is empty, maybe we'd want to know via a log, but I don't want Crash and Perf to have to handle nil cases. WDYT?

Reply:
I would prefer to send NIL and handle the case in the product SDKs. I would want to be able to differentiate situations where a sessionID is valid and invalid. With empty it takes it to an unpredictable scenario.

Aside: Even if empty, we might have to do a check on the session ID length in the product SDKs. Isn't that true? So, instead of empty check, we could as well do a NIL check?

Reply:
Sounds good - just updated that

Reply:
Thanks for updating it.